### PR TITLE
New engine.js first draft.

### PR DESCRIPTION
**Pull request by madrang**

### CODE REVIEW

**Comment by madrang** [comment]
The engine.js unit testing plugin is here https://github.com/madrang/sd-ui-plugins

**Review by rbertus2000 — APPROVED** [review]
Madrang and I tested and debugged a lot. Should run stable now without bigger issues.

**Comment by cmdr2** [comment]
Doesn't this also need some backend changes, specifically to the `task_cache` to allow concurrent tasks to work without throwing 'task id is not the same' errors on `GET /image/stream/session-id/task-id`?

File: ui/media/js/engine.js [hunk 1]
```diff
@@ -0,0 +1,1264 @@
...
+    const EVENT_UNHANDLED_REJECTION = 'unhandledRejection'
+    const EVENT_TASK_QUEUED = 'taskQueued'
+    const EVENT_TASK_START = 'taskStart'
+    const EVENT_TASK_END = 'taskEnd'
+    const EVENT_TASK_ERROR = 'task_error'
+    const EVENT_UNEXPECTED_RESPONSE = 'unexpectedResponse'
+    const EVENTS_TYPES = [
+        EVENT_IDLE
+        , EVENT_STATUS_CHANGED
+        , EVENT_UNHANDLED_REJECTION
```
[the remark this answers is not in the document]
Comment (madrang) [reply ui/media/js/engine.js:187]: Alright, I'll change it. It's just that comma first is more safe for git, so not too long ago we moved to this at work.
It leaves not trailing comma, remove the needs to edit two lines when adding only one and make it very obvious when a comma is missing.

**Comment by madrang** [comment]
The backend changes are needed to make it run more than one concurrent render.
But the frontend is now ready for it. But yes at the moment there is a single hardcoded 1 somewhere that keep it from queuing more than on task until the backend changes are made.

**Comment by madrang** [comment]
Did the changes for the end of line comma.

**Comment by madrang** [comment]
Even if the backend changes for concurrency aren't done yet, it still add task reordering and plugin control and usage for engine.js

Later we can update the line to enable concurrency whenever the backend is ready and having the UI ready it will help me test the backend before I make the PR.

**Pull request merged**